pkg/controller/operators/catalog/operator.go

```diff
@@ -1056,6 +1057,19 @@ func (o *Operator) ExecutePlan(plan *v1alpha1.InstallPlan) error {
 				// Attempt to create the CRD.
 				_, err = o.OpClient.ApiextensionsV1beta1Interface().ApiextensionsV1beta1().CustomResourceDefinitions().Create(&crd)
 				if k8serrors.IsAlreadyExists(err) {
+					currentCRD, _ := o.OpClient.ApiextensionsV1beta1Interface().ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.GetName(), metav1.GetOptions{})
+					// Compare 2 CRDs to see if it needs to be updatetd
+					if !reflect.DeepEqual(crd, *currentCRD) {
+						// Verify CRD ownership, only attempt to update if
+						// CRD has only one owner
+						if len(existingCRDOwners[currentCRD.GetName()]) == 1 {
```

Comment:
Unfortunately, existingCRDOwners only tells you about owners in the same namespace:
https://github.com/operator-framework/operator-lifecycle-manager/blob/07dd9447712d5d3d21975f3275c9fa51185a29de/pkg/controller/operators/catalog/operator.go#L1399-L1420

For upgrade safety for CRDs, we need to be sure that there's no owner in _all_ namespaces.

The simplest thing to do would be to have a similar function but which collects from all namespace.

But, the nicest way to do this would be to add an extra IndexFunc to our CSV cache to key off of provided APIs. If we do that, then this becomes a very simple query into the cache. We're already storing all CSVs in the cache so adding the extra index should be very low overhead.

Reply:
@ecordell Is this what you had in mind? I'm not completely sure if the indexes are correct.

```
	sharedInformerFactory := externalversions.NewSharedInformerFactory(crClient, wakeupInterval)
	csvInformer := sharedInformerFactory.Operators().V1alpha1().ClusterServiceVersions()
	op.lister.OperatorsV1alpha1().RegisterClusterServiceVersionLister(metav1.NamespaceAll, csvInformer.Lister())
	csvInformer.Informer().AddIndexers(cache.Indexers{"providedAPIs": func(obj interface{}) ([]string, error) {
		indicies := []string{}

		csv, ok := obj.(*v1alpha1.ClusterServiceVersion)
		if !ok {
			return indicies, fmt.Errorf("invalid object of type: %T", obj)
		}

		for _, crd := range csv.Spec.CustomResourceDefinitions.Owned {
			indicies = append(indicies, fmt.Sprintf("%s=%s", crd.Name, csv.GetName()))
		}
		for _, api := range csv.Spec.APIServiceDefinitions.Owned {
			indicies = append(indicies, fmt.Sprintf("%s=%s", api.Group, csv.GetName()))
		}

		return indicies, nil
	}})
	op.csvProvidedAPIIndexer = csvInformer.Informer().GetIndexer()
```

Reply:
I just realized that the code here is in the catalog operator. Given that, I think it's probably best to just extend getExistingApiOwners as necessary (or create something new).

Reply:
@jpeeler Yep, something like that! I think the current way the informers are set up you may need one per namespace, IIRC (that may change with @njhale 's prs)

The index itself should be something we can query:

```go
fmt.Sprintf("provided=%s/%s/%k", csvGroup, csvVersion, csvKind);
```
That way you can get the list of CSVs that provide an api with `provided=database.coreos.com/v1alpha1/EtcdCluster`, etc